squad/api/rest.py

```diff
@@ -503,53 +550,66 @@ def email(self, request, pk=None):
-            pr_status = self.get_object().status
-            notification = Notification(pr_status, baseline)
-            produce_html = self.get_object().project.html_mail
-            if output_format == "text/html":
-                produce_html = True
-            try:
-                txt, html = notification.message(produce_html, template)
-                if len(html) > 0:
-                    return HttpResponse(html, content_type=output_format)
-                return HttpResponse(txt, content_type=output_format)
-            except TemplateSyntaxError as e:
-                data = {
-                    "lineno": e.lineno,
-                    "message": e.message
-                }
-                if template is not None:
-                    data.update({
-                        "txt": template.plain_text,
-                        "html": template.html
-                    })
-                return Response(data, status=status.HTTP_500_INTERNAL_SERVER_ERROR)
-            except TypeError as te:
-                data = {"message": str(te)}
-                if template is not None:
-                    data.update({
-                        "txt": template.plain_text,
-                        "html": template.html
-                    })
-                return Response(data, status=status.HTTP_500_INTERNAL_SERVER_ERROR)
+                # return created=False to avoid running prepare_report
+                return save_delayed_report(delayed_report, data, status.HTTP_500_INTERNAL_SERVER_ERROR), False
+
+        return delayed_report, created
+
+    @detail_route(methods=['get'], suffix='email')
+    def email(self, request, pk=None):
+        """
+        This method produces the body of email notification for the build.
+        By default it uses the project settings for HTML and template.
+        These settings can be overwritten by using GET parameters:
+         * output - sets the output format (text/plan, text/html)
+         * template - sets the template used (id of existing template or
+                      "default" for default SQUAD templates)
+         * force - force email report re-generation even if there is
+                   existing one cached
+        """
+        force = request.query_params.get("force", False)
+        delayed_report, created = self.__return_delayed_report(request)
+        if created or force:
+            delayed_report = prepare_report(delayed_report.pk)
+        if delayed_report.status_code != status.HTTP_200_OK:
+            return Response(yaml.load(delayed_report.error_message), delayed_report.status_code)
+        if delayed_report.output_html:
+            return HttpResponse(delayed_report.output_html, content_type=delayed_report.output_format)
+        return HttpResponse(delayed_report.output_text, content_type=delayed_report.output_format)
+
+    @detail_route(methods=['get', 'post'], suffix='report', permission_classes=[AllowAny])
+    def report(self, request, pk=None):
```

Comment:
no. DelayedReport is created in line 605. If you call the same URL immediately after it will trigger the notifications. Honestly I'm not sure why I did that. It doesn't look right. Notifications should only be sent when the report is ready. So the whole 'else' block should be removed (I think).

In line 607 the DelayedReport is sent for processing. Before that the status_code field is empty. After prepare_report finishes, status_code should be filled in with some value. This way client knows when the report is ready (I forgot to add this line to the docs).

Reply:
I think the else block could be removed, or maybe just return DelayedReport status, because the notifications are already sent after report finished processing. What do you think?

Reply:
You're right. I'll make changes